File: frontend/src/types/merge-queue.ts
// Copyright (c) 2026 Valaris Studio
// SPDX-License-Identifier: AGPL-3.0-or-later

// Mirrors `MergeQueueEntryRead` (backend/app/schemas/merge_queue.py) and the
// closed state enum in backend/app/models/agents/merge_queue.py.
export type MergeQueueState =
  | "queued"
  | "merging"
  | "merged"
  | "conflict"
  | "failed"
  | "blocked_pending_consolidation";

export interface MergeQueueEntry {
  id: string;
  repo_id: string;
  integration_branch: string;
  card_id: string;
  pr_url: string;
  pr_branch: string;
  workspace_id: string;
  enqueued_at: string;
  // When the entry FIRST entered the queue. `enqueued_at` is the FIFO position
  // and gets bumped to the back on every retry, so only this is a usable
  // stuck-duration clock. Null for rows predating backend migration 088.
  first_enqueued_at?: string | null;
  state: MergeQueueState;
  attempt_count: number;
  error_message?: string | null;
  merged_at?: string | null;
}

File: backend/app/schemas/merge_queue.py
# Copyright (c) 2026 Valaris Studio
# SPDX-License-Identifier: AGPL-3.0-or-later

import uuid
from datetime import datetime

from pydantic import field_validator

from app.core.json_response import UTCModel
from app.schemas.git.git_repo import validate_branch_name


class MergeQueueEnqueueRequest(UTCModel):
    card_id: uuid.UUID
    repo_id: uuid.UUID
    pr_url: str
    pr_branch: str
    # Optional: when omitted, the service falls back to repo.integration_branch.
    integration_branch: str | None = None

    # These take precedence over the repo's own branches and reach `git push` /
    # `git merge` as bare argv, where `--receive-pack=cmd` makes git spawn
    # `/bin/sh -c cmd`. Same validator as the repo schema — this is the second
    # ingress for a branch name, not a different kind of value.
    @field_validator("pr_branch", "integration_branch")
    @classmethod
    def _check_branch(cls, v: str | None) -> str | None:
        return validate_branch_name(v)


class MergeQueueReEnqueueRequest(UTCModel):
    """PAR-3c: re-queue an existing entry by its parent card_id.

    The consolidator pipeline calls this once it has resolved the conflict;
    the worker then re-attempts the merge. No repo / pr fields — the entry
    already has them.
    """

    card_id: uuid.UUID


class MergeQueueEntryRead(UTCModel):
    id: uuid.UUID
    repo_id: uuid.UUID
    integration_branch: str
    card_id: uuid.UUID
    pr_url: str
    pr_branch: str
    workspace_id: uuid.UUID
    enqueued_at: datetime
    # The FIFO position (`enqueued_at`) is bumped to the back of the queue on
    # every retry, so clients cannot derive stuck-duration from it. This is the
    # origin clock. NULL for rows predating migration 088.
    first_enqueued_at: datetime | None = None
    state: str
    attempt_count: int
    error_message: str | None = None
    merged_at: datetime | None = None

    model_config = {"from_attributes": True}
